motionplan/rrtStarConnect.go

```diff
@@ -176,64 +225,165 @@ func (mp *rrtStarConnectMotionPlanner) rrtBackgroundRunner(ctx context.Context,
+	// 4) further iterations change our best node by close-to-zero amounts
+	// 5) we have iterated more than maxExtendIter times
+	for i := 0; i < maxExtendIter; i++ {
+		select {
+		case <-ctx.Done():
+			mchan <- oldNear
+			return
+		default:
+		}
+
+		dist := mp.planOpts.DistanceFunc(&Segment{StartConfiguration: near.Q(), EndConfiguration: target.Q()})
+		oldDist := mp.planOpts.DistanceFunc(&Segment{StartConfiguration: oldNear.Q(), EndConfiguration: target.Q()})
+		switch {
+		case dist < mp.algOpts.JointSolveDist:
+			mchan <- near
+			return
+		case dist > oldDist:
+			mchan <- oldNear
+			return
+		}
+
+		oldNear = near
+		newNear := make([]referenceframe.Input, 0, len(near.Q()))
+
+		// alter near to be closer to target
+		for j, nearInput := range near.Q() {
+			if nearInput.Value == target.Q()[j].Value {
+				newNear = append(newNear, nearInput)
+			} else {
+				v1, v2 := nearInput.Value, target.Q()[j].Value
+				newVal := math.Min(qstep[j], math.Abs(v2-v1))
+				// get correct sign
+				newVal *= (v2 - v1) / math.Abs(v2-v1)
+				newNear = append(newNear, referenceframe.Input{nearInput.Value + newVal})
+			}
+		}
+		// Check whether oldNear -> newNear path is a valid segment, and if not then set to nil
+		newNear = mp.constrainNear(oldNear.Q(), newNear)
+
+		if newNear != nil {
+			nearDist := mp.planOpts.DistanceFunc(&Segment{StartConfiguration: oldNear.Q(), EndConfiguration: newNear})
+			if nearDist < math.Pow(mp.algOpts.JointSolveDist, 3) {
+				if !doubled {
+					doubled = true
+					// Check if doubling qstep will allow escape from the identical configuration
+					// If not, we terminate and return.
+					// If so, qstep will be reset to its original value after the rescue.
+					for i, q := range qstep {
+						qstep[i] = q * 2.0
+					}
+					continue
+				} else {
+					// We've arrived back at very nearly the same configuration again; stop solving and send back oldNear.
+					// Do not add the near-identical configuration to the RRT map
+					mchan <- oldNear
+					return
+				}
+			}
+			if doubled {
+				copy(qstep, mp.algOpts.qstep)
+				doubled = false
+			}
+
+			neighbors := kNearestNeighbors(mp.planOpts, rrtMap, &basicNode{q: newNear}, mp.algOpts.NeighborhoodSize)
+
+			near = &basicNode{q: newNear, cost: neighbors[0].node.Cost() + neighbors[0].dist}
+			rrtMap[near] = oldNear
+
+			// rewire the tree
+			for i, thisNeighbor := range neighbors {
+				// dont need to try to rewire nearest neighbor, so skip it
+				if i == 0 {
+					continue
+				}
+
+				// check to see if a shortcut is possible, and rewire the node if it is
+				connectionCost := mp.planOpts.DistanceFunc(&Segment{
+					StartConfiguration: thisNeighbor.node.Q(),
+					EndConfiguration:   near.Q(),
+				})
+				cost := connectionCost + near.Cost()
+
+				// If 1) we have a lower cost, and 2) the putative updated path is valid
+				if cost < thisNeighbor.node.Cost() && mp.checkPath(target.Q(), thisNeighbor.node.Q()) {
```

Comment:
We've got `mp.checkPath`, and we have `constrainNear`, which does a `Transform` and then a `mp.planOpts.CheckSegmentAndStateValidity` call.

`mp.checkPath` does a `mp.planOpts.CheckSegmentAndStateValidity` as well, but without the intermediate transform.

We should be consistent about which one we call. Since constraints can and will fill in the Transform if needed, it should be ok to just use `checkPath` everywhere and drop the contents of `constrainNear` entirely